### PR TITLE
Wrapper generator

### PR DESCRIPTION
**Pull request by ptyszko**
* Add support for namespace files, same format as [Libamtrack's NAMESPACE](https://github.com/libamtrack/library/blob/master/distributions/R/NAMESPACE)
* Refactor wrapper generation in order to make creating different kind of wrappers easier

### CODE REVIEW

File: scripts/header_parser.py [hunk 1]
```diff
@@ -90,112 +154,113 @@ def __init__(self,
...
-    :type fun: CppMethod
+    ::
+        some.func <- function(param2){
+          param1 <- max(length(param2))
+          param2 <- as.integer(param2)
+          param3 <- vector(mode="integer", length=param1)
+          AUTO_RET_PARAMS <- .C("some_func", param1, param2, param3)
+          param2 <- AUTO_RET_PARAMS$param2
+          param3 <- AUTO_RET_PARAMS$param3
+          AUTO_RETVAL <- list("param2" = param2, "param3" = param3
```
Comment (grzanka) [inline scripts/header_parser.py:222]: Missing `)` ?

File: scripts/header_parser.py [hunk 2]
```diff
@@ -90,112 +154,113 @@ def __init__(self,
...
-    int some_func(const int *param1, int *param2, float *param3);
-    ```
+    ::
+        /**
+         * @brief description
+         * @param[in] param1
+         * @param[in,out] param2 array of length param1
+         * @param[out] param3 array of length param1
+         */
+        int some_func(const int *param1, int *param2, float *param3);
```
Comment (grzanka) [inline scripts/header_parser.py:211]: Does it make sense to use a pointer `const int *param1` as array length ?

File: scripts/header_parser.py [hunk 3]
```diff
@@ -90,112 +154,113 @@ def __init__(self,
...
-      AUTO_RETVAL <- list("param2" = param2, "param3" = param3
-      return(AUTO_RETVAL)
-    }
-    ```
-
-    :param fun: The function signature together with doxygen comment
-    :type fun: CppMethod
+    ::
+        some.func <- function(param2){
+          param1 <- max(length(param2))
```
Comment (grzanka) [inline scripts/header_parser.py:216]: Why max ?

File: scripts/header_parser.py [hunk 4]
```diff
@@ -90,112 +154,113 @@ def __init__(self,
...
+          param2 <- as.integer(param2)
+          param3 <- vector(mode="integer", length=param1)
```
Comment (grzanka) [inline scripts/header_parser.py:218]: Why integer mode ?

File: scripts/header_parser.py [hunk 5]
```diff
@@ -246,19 +311,80 @@ def create_wrapper_for_function(fun: CppMethod):
...
 
-def create_wrappers_for_header_file(path: str, out_dir: str):
+def create_wrappers_for_header_file(path: str,
+                                    out_dir: str,
+                                    namespace: Collection[str] = None):
     Path(out_dir).mkdir(parents=True, exist_ok=True)
-    out_path = Path(out_dir + '/' + Path(path).name.replace('.h', '.R'))
-    with open(out_path, 'w') as fout:
-        for func in extract_functions_from_file(path):
+    r_out_path = Path(out_dir + '/' + Path(path).name.replace('.h', '.R'))
```
Comment (grzanka) [inline scripts/header_parser.py:321]: Following should also work and be more OS-agnostic:
```
r_out_path = Path(out_dir, Path(path).name.replace('.h', '.R'))
```

File: scripts/header_parser.py [hunk 6]
```diff
@@ -246,19 +311,80 @@ def create_wrapper_for_function(fun: CppMethod):
...
+
+    :param path: path to your namespace file
+    :return:
+    """
+    try:
+        # in theory we don't expect any non-ascii characters, but better safe than sorry
+        with open(path, 'r', encoding='utf-8') as fin:
+            names = fin.readlines()
+    except IOError:
+        print("Couldn't retrieve namespace, assuming all functions")
```
Comment (grzanka) [inline scripts/header_parser.py:374]: Better to add (absolute?) path of the file for which retrieving failed.

**Review by grzanka — CHANGES_REQUESTED** [review]

**Comment by grzanka** [comment]
@ptyszko please fix failing test

File: scripts/header_parser.py (hunk 3 above)
Comment (ptyszko) [reply scripts/header_parser.py]: That was my mistake when making an example, already fixed

File: scripts/header_parser.py (hunk 4 above)
Comment (ptyszko) [reply scripts/header_parser.py]: That was my mistake when making an example, already fixed

File: scripts/header_parser.py (hunk 1 above)
Comment (ptyszko) [reply scripts/header_parser.py]: That was my mistake when making an example, already fixed

**Review by grzanka — APPROVED** [review]
Looks OK

**Pull request merged**